### PR TITLE
GH-119500: Remove slice opcodes and handle constant slices in the compiler.

### PR DESCRIPTION
**Pull request by markshannon**
This PR:
* Removes the `BINARY_SLICE` and `STORE_SLICE` opcodes
* Converts constant slices to constants in the AST optimizer.
* Merges the free-threading and normal code comparison functions, to handle equal, but not identical code objects.

* Issue: gh-119500

### CODE REVIEW

File: Objects/codeobject.c [hunk 1]
```diff
@@ -1904,25 +1904,92 @@ code_repr(PyCodeObject *co)
...
+        return PyObject_RichCompareBool(op1, op2, Py_EQ);
+    }
+    if (PyTuple_CheckExact(op1)) {
+        Py_ssize_t size = PyTuple_GET_SIZE(op1);
+        if (size != PyTuple_GET_SIZE(op2)) {
+            return 0;
+        }
+        for (Py_ssize_t i = 0; i < size; i++) {
+            if (compare_constants(PyTuple_GET_ITEM(op1, i),
+                PyTuple_GET_ITEM(op2, i)) == 0) {
```
Comment (iritkatriel) [inline Objects/codeobject.c:1932]:
```suggestion
            if (compare_constants(PyTuple_GET_ITEM(op1, i),
                                  PyTuple_GET_ITEM(op2, i)) == 0) {
```

File: Python/marshal.c [hunk 2]
```diff
@@ -1525,6 +1534,31 @@ r_object(RFILE *p)
...
+        stop = r_object(p);
+        if (stop == NULL) {
+            goto cleanup;
+        }
+        step = r_object(p);
+        if (step == NULL) {
+            goto cleanup;
+        }
+        retval = PySlice_New(start, stop, step);
+        assert(retval != NULL);
```
Comment (iritkatriel) [inline Python/marshal.c:1554]: assert? Not memory error?

File: Python/marshal.c [hunk 3]
```diff
@@ -1484,6 +1492,7 @@ r_object(RFILE *p)
 
             v = (PyObject *)_PyCode_New(&con);
             if (v == NULL) {
+                assert(0);
```
Comment (iritkatriel) [inline Python/marshal.c:1495]: ?

**Comment by markshannon** [comment]
Handling constant slices might be worthwhile, but it looks like we are better off keeping `BINARY_SLICE` for now.

**Pull request closed without merging**